### ISSUE TITLE
code question for compute_loss in ppo_trainer

### ISSUE THREAD

**Issue by Yuhuajoe** [tatsu-lab/alpaca_farm#52]
I'm wondering that  we have achieved policy's outputs in rollouts, why we use self.policy again during compute_loss module but not reuse outputs in rollouts?
is there something different for two self.policy or some other reason?

code review are as belows: 
code for compute_loss
```
        values, old_logprob, returns, advantages, queries, query_attn_masks, responses = common.prepare_inputs(
            common.unpack_dict(
                rollouts,
                keys=("values", "logprobs", "returns", "advantages", "queries", "query_attn_masks", "responses"),
            ),
            device=self.accelerator.device,
        )
        outputs = self.policy(queries, query_attn_masks, responses, temperature=self.args.temperature)

        vpred = outputs["values"]
        vpredclipped = torch.clamp(
            vpred,
            min=values - self.args.cliprange_value,
            max=values + self.args.cliprange_value,
        )
```

code for rollouts_batch, subset of rollouts
```
           rollouts_batch = {"queries": queries, "query_attn_masks": query_attn_masks, "responses": responses}
            policy_outputs = self.policy(**rollouts_batch, temperature=self.args.temperature)
            policy_outputs = common.unpack_dict(
                policy_outputs, keys=("logprobs", "values", "entropies"), return_type=dict
            )
            rollouts_batch.update(policy_outputs)
```

**Comment by lxuechen** [comment]
You need to redo the forward pass for `self.policy`, since the parameters may have changed compared to when the rollout was obtained. Ultimately, this is because PPO performs multiple gradient updates on the set of rollouts obtained every once in a while.

**Comment by Yuhuajoe** [comment]
We obtain rollouts with a big batch size( 512) first, but do gradient updates with a small batch size(16), so we need to redo self.policy during compute_loss.
Is this right?

**Comment by lxuechen** [comment]
It's not only about the batch size, but also multiple epochs.

**Comment by Yuhuajoe** [comment]
thx for your reply, i got it.
